### ISSUE TITLE
compilation error: use of undeclared identifier (macosx)

### ISSUE THREAD

**Issue by jji2019** [bytedeco/javacpp-presets#799]
Hi.
I followed the build instruction https://github.com/bytedeco/javacpp-presets/wiki/Build-Environments, but still got compilation error like below.
Please share if any additional suggestion. Appreciate that.

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] JavaCPP Presets 1.5.2-SNAPSHOT ..................... SUCCESS [  0.716 s]
[INFO] JavaCPP Presets for OpenBLAS 0.3.7-1.5.2-SNAPSHOT .. SUCCESS [10:37 min]
[INFO] JavaCPP Presets for OpenCV 4.1.1-1.5.2-SNAPSHOT .... SUCCESS [13:48 min]
[INFO] JavaCPP Presets for FFmpeg 4.2.1-1.5.2-SNAPSHOT .... SUCCESS [18:06 min]
[INFO] JavaCPP Presets for FlyCapture 2.13.3.31-1.5.2-SNAPSHOT FAILURE [  1.119 s]

> /Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:294:99: **error: use of undeclared identifier**
      'MultiSyncLibrary'
static void JavaCPP_org_bytedeco_flycapture_FlyCapture2_SyncManager_deallocate(void *p) { delete (MultiSyncLibrary::SyncManager*)p; }
                                                                                                  ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:294:129: **error: expected expression**
static void JavaCPP_org_bytedeco_flycapture_FlyCapture2_SyncManager_deallocate(void *p) { delete (MultiSyncLibrary::SyncManager*)p; }
                                                                                                                                ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:295:106: **error: use of undeclared identifier**
      'MultiSyncLibrary'
static void JavaCPP_org_bytedeco_flycapture_FlyCapture2_SyncManager_deallocateArray(void* p) { delete[] (MultiSyncLibrary::SyncManager*)p; }
                                                                                                         ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:295:136: **error: expected expression**
static void JavaCPP_org_bytedeco_flycapture_FlyCapture2_SyncManager_deallocateArray(void* p) { delete[] (MultiSyncLibrary::SyncManager*)p; }
                                                                                                                                       ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:314:18: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        { sizeof(MultiSyncLibrary::SyncManager) },
                 ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:445:9: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        MultiSyncLibrary::SyncManager* rptr = new MultiSyncLibrary::SyncManager[arg0];
        ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:445:51: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        MultiSyncLibrary::SyncManager* rptr = new MultiSyncLibrary::SyncManager[arg0];
                                                  ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:457:5: **error: use of undeclared identifier**
      'MultiSyncLibrary'
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
    ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:457:43: **error: use of undeclared identifier**
      'MultiSyncLibrary'
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
                                          ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:457:73: **error: expected expression**
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
                                                                        ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:467:9: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        MultiSyncLibrary::PGRSyncError rval = (MultiSyncLibrary::PGRSyncError)ptr->RescanMasterTimingBus();
        ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:467:48: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        MultiSyncLibrary::PGRSyncError rval = (MultiSyncLibrary::PGRSyncError)ptr->RescanMasterTimingBus();
                                               ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:479:5: **error: use of undeclared identifier**
      'MultiSyncLibrary'
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
    ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:479:43: **error: use of undeclared identifier**
      'MultiSyncLibrary'
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
                                          ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:479:73: **error: expected expression**
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
                                                                        ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:489:9: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        MultiSyncLibrary::PGRSyncMessage rval = (MultiSyncLibrary::PGRSyncMessage)ptr->GetSyncStatus();
        ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:489:50: **error: use of undeclared identifier**
      'MultiSyncLibrary'
        MultiSyncLibrary::PGRSyncMessage rval = (MultiSyncLibrary::PGRSyncMessage)ptr->GetSyncStatus();
                                                 ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:501:5: **error: use of undeclared identifier**
      'MultiSyncLibrary'
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
    ^
/Users/my/javacpp-presets/flycapture/target/native/org/bytedeco/flycapture/macosx-x86_64/jniFlyCapture2.cpp:501:43: **error: use of undeclared identifier**
      'MultiSyncLibrary'
    MultiSyncLibrary::SyncManager* ptr = (MultiSyncLibrary::SyncManager*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
                                          ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]

btw, since there is no ngraph for macosx https://repo1.maven.org/maven2/org/bytedeco/ngraph/0.22.0-1.5.1/, it is required to compile the native libraries manually for macosx. Is there any suggestion to build only ngraph project from Mac OS?

Below is what I tried to build single sub project "ngraph". With modification in ngraph/cppbuild.sh

>  53 #cp onnx/bin/libonnxifi.so ../lib/
>  54 cp onnx/bin/libonnxifi.dylib ../lib/

I got below error:
> /Users/my/javacpp-presets/ngraph/target/native/org/bytedeco/ngraph/macosx-x86_64/jningraph.cpp:3023:58: error: cannot initialize a parameter of type
      '::onnxBackendID *' (aka 'void **') with an lvalue of type 'void *'
        ngraph::onnxifi::BackendManager::get_backend_ids(ptr0, (std::size_t*)ptr1);
                                                         ^~~~
/Users/my/javacpp-presets/ngraph/cppbuild/macosx-x86_64/include/ngraph/frontend/onnxifi/backend_manager.hpp:41:58: note: passing argument to parameter 'backend_ids'
      here
            static void get_backend_ids(::onnxBackendID* backend_ids, std::size_t* count)

**Comment by saudet** [comment]
FlyCapture isn't supported on Mac.

**Comment by jji2019** [comment]
@saudet  very appreciate your reply. Apart from "FlyCapture", is nGraph supported on Mac OSX?

**Comment by saudet** [comment]
Not yet, but it shouldn't be hard to make that one work on Mac.

**Comment by jji2019** [comment]
Thanks for the information. :D

**Comment by EmergentOrder** [comment]
Sorry, I can't / don't support Mac (or Windows for that matter), I only have a Linux box.

**Comment by saudet** [comment]
I got it working, just had to change a couple of lines actually. I'll push
that later today.

**Comment by saudet** [comment]
Ok, it's done! Let me know if you encounter any issues with the snapshots: http://bytedeco.org/builds/

**Comment by jji2019** [comment]
@saudet: Thanks for the information. Since nGraph needs mkl-dnn to be compiled first, I got the errors while doing the compilation in mkl-dnn sub project. ps. I am using macOS Catalina

`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/resource.h:443:34: error: expected initializer before '__OSX_AVAILABLE_STARTING'`

and 

`/Users/my/javacpp-presets-master/mkl-dnn/cppbuild/macosx-x86_64/mkl-dnn-0.21.1/src/cpu/xbyak/xbyak.h:317:12: error: 'posix_memalign' was not declared in this scope`

**Comment by saudet** [comment]
Why are you trying to compile them? Are you having problems with the
precompiled binaries?

**Comment by jji2019** [comment]
I was experiencing that the pre-compiled nd4j related *.dylib was loaded without problem to macOS Mojave but is loaded with exception in macOS Catalina. So I would like to try compiling from my latest OS. Btw, I could also try the pre-compiled nGraph. Will update my status once I got it work. Really appreciate your help.

**Comment by saudet** [comment]
Ok, I see. Both nGraph and MKL-DNN have been upgraded yesterday, so please try again!

**Comment by jji2019** [comment]
Hi,

I am able to compile ngraph project latest source code (pull from master on Oct21, 2019) by the command: `mvn clean install -rf :ngraph -U` which downloads pre-compiled jar if any prerequisite dependencies.
OS: `MacOS Catalina 10.15`
Xcode: `11.1`

But the latest version of ngraph platform `ngraph-platform 0.26.0-1.5.2-SNAPSHOT` cannot load CPU backend. The output stuck at `loading cpu backend...` with below sample code from https://github.com/bytedeco/javacpp-presets/blob/master/ngraph/README.md
`System.out.println("loading cpu backend...");`
`Backend backend = Backend.create("CPU");`
`Shape shape = new Shape(new SizeTVector(1, 2));`
`System.out.println("creating input tensor...");`
`Tensor input = backend.create_tensor(org.bytedeco.ngraph.global.ngraph.f32(), shape);`

**Comment by saudet** [comment]
If you set the "org.bytedeco.javacpp.logger.debug" system property to "true", do you get anything more on the console?

**Comment by jji2019** [comment]
Here is the result with debug on `-Dorg.bytedeco.javacpp.logger.debug=true`

```
loading cpu backend...
Loading class org.bytedeco.ngraph.global.ngraph
Loading class org.bytedeco.ngraph.global.ngraph
Loading /Users/my/.javacpp/cache/mkl-dnn-0.21.2-1.5.2-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/mkldnn/macosx-x86_64/libiomp5.dylib
Loading /Users/my/.javacpp/cache/mkl-dnn-0.21.2-1.5.2-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/mkldnn/macosx-x86_64/libmklml.dylib
Loading library mklml_intel
Failed to load for mklml_intel: java.lang.UnsatisfiedLinkError: no mklml_intel in java.library.path
Loading /Users/my/.javacpp/cache/ngraph-0.26.0-1.5.2-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/ngraph/macosx-x86_64/libmkldnn.dylib
Loading /Users/my/.javacpp/cache/ngraph-0.26.0-1.5.2-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/ngraph/macosx-x86_64/libncurses.6.dylib
Loading /Users/my/.javacpp/cache/ngraph-0.26.0-1.5.2-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/ngraph/macosx-x86_64/libonnxifi.dylib
Loading /Users/my/.javacpp/cache/ngraph-0.26.0-1.5.2-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/ngraph/macosx-x86_64/libngraph.dylib
```

**Comment by saudet** [comment]
Looks alright, so it's a problem with nGraph. Try to put its log in debug
mode too.

**Comment by jji2019** [comment]
Sorry for late reply. I hope I knew how to enable the debug mode for nGraph.

**Comment by saudet** [comment]
ONNX Runtime appears to have issues running under the JVM, yes, but that's probably not related.

For nGraph, it looks like we can get more information by setting the NGRAPH_CPU_DEBUG_TRACER environment variable to 1:
https://github.com/NervanaSystems/ngraph/blob/master/doc/sphinx/source/frameworks/other.rst

**Comment by saudet** [comment]
In any case, working binaries for at least nGraph have been released with version 1.5.2. Enjoy!

**Comment by saudet** [comment]
ONNX Runtime 1.0.0 is also running fine now:
https://github.com/bytedeco/javacpp-presets/tree/master/onnxruntime/
